sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java

```diff
@@ -411,4 +327,372 @@ public void run() {
       }
     }
   }
+
+  @Immutable
+  @AutoValue
+  public abstract static class Config {
+
+    private static final long DEFAULT_SCHEDULE_DELAY_MILLIS = 5000;
+    private static final int DEFAULT_MAX_QUEUE_SIZE = 2048;
+    private static final int DEFAULT_MAX_EXPORT_BATCH_SIZE = 512;
+    private static final int DEFAULT_EXPORT_TIMEOUT_MILLIS = 30_000;
+    private static final boolean DEFAULT_EXPORT_ONLY_SAMPLED = true;
+
+    public abstract boolean isExportOnlySampled();
+
+    public abstract long getScheduleDelayMillis();
+
+    public abstract int getMaxQueueSize();
+
+    public abstract int getMaxExportBatchSize();
+
+    public abstract int getExporterTimeoutMillis();
+
+    /**
+     * Creates a {@link Config} object using the default configuration.
+     *
+     * @return The {@link Config} object.
+     * @since 0.4.0
+     */
+    public static Config getDefault() {
+      return builder().build();
+    }
+
+    /**
+     * Creates a {@link Config} object reading the configuration values from the environment and
+     * from system properties. System properties override values defined in the environment. If a
+     * configuration value is missing, it uses the default value.
+     *
+     * @return The {@link Config} object.
+     * @since 0.4.0
+     */
+    public static Config loadFromDefaultSources() {
+      return builder().readEnvironment().readSystemProperties().build();
+    }
+
+    /**
+     * Returns a new {@link Builder} with default options.
+     *
+     * @return a new {@code Builder} with default options.
+     * @since 0.4.0
+     */
+    public static Builder builder() {
+      return new AutoValue_BatchSpansProcessor_Config.Builder()
+          .setScheduleDelayMillis(DEFAULT_SCHEDULE_DELAY_MILLIS)
+          .setMaxQueueSize(DEFAULT_MAX_QUEUE_SIZE)
+          .setMaxExportBatchSize(DEFAULT_MAX_EXPORT_BATCH_SIZE)
+          .setExporterTimeoutMillis(DEFAULT_EXPORT_TIMEOUT_MILLIS)
+          .setExportOnlySampled(DEFAULT_EXPORT_ONLY_SAMPLED);
+    }
+
+    @AutoValue.Builder
+    public abstract static class Builder {
+
+      private static final String KEY_SCHEDULE_DELAY_MILLIS = "otel.bsp.schedule.delay";
+      private static final String KEY_MAX_QUEUE_SIZE = "otel.bsp.max.queue";
+      private static final String KEY_MAX_EXPORT_BATCH_SIZE = "otel.bsp.max.export.batch";
+      private static final String KEY_EXPORT_TIMEOUT_MILLIS = "otel.bsp.export.timeout";
+      private static final String KEY_SAMPLED = "otel.bsp.export.sampled";
+
+      @VisibleForTesting
+      protected enum NamingConvention {
+        DOT {
+          @Nonnull
+          @Override
+          public String normalize(@Nonnull String key) {
+            return key.toLowerCase();
+          }
+        },
+        ENV_VAR {
+          @Nonnull
+          @Override
+          public String normalize(@Nonnull String key) {
+            return key.toLowerCase().replace("_", ".");
+          }
+        };
+
+        @Nonnull
```

Comment:
We don't use the `@Nonnull` convention. Please avoid adding this new convention yet.